File: Solucion_Codigo/Proyecto Bimestral Python/modelo/metodos.py
class Metodos:
    CATEGORIAS = ["vivienda", "educacion", "alimentacion", "vestimenta", "salud", "turismo"]

    def __init__(self, nombre, cedula, direccion):
        self.nombre = nombre
        self.cedula = cedula
        self.direccion = direccion
        self.sueldos_mensuales = [0.0] * 12
        self.facturas_por_categoria = [0.0] * len(self.CATEGORIAS)
        self.total_ingresos = 0.0
        self.total_deducciones = 0.0
        self.impuestos_calculados = 0.0

    def agregar_sueldo(self, mes, monto):
        if 0 <= mes < 12 and monto > 0:
            self.sueldos_mensuales[mes] = monto

    def agregar_factura(self, monto, categoria):
        try:
            index = self.CATEGORIAS.index(categoria.lower())
            self.facturas_por_categoria[index] += monto
        except ValueError:
            pass

    def calcular_total_ingresos(self):
        self.total_ingresos = sum(self.sueldos_mensuales)

    def calcular_deducciones(self):
        total_facturas = sum(self.facturas_por_categoria)
        limite = min(14754.00, self.total_ingresos * 0.5)
        self.total_deducciones = min(total_facturas, limite)

    def calcular_impuestos(self):
        base = self.total_ingresos - self.total_deducciones
        if base <= 11722:
            self.impuestos_calculados = 0
        elif base <= 14930:
            self.impuestos_calculados = (base - 11722) * 0.05
        elif base <= 18660:
            self.impuestos_calculados = (base - 14930) * 0.10 + 160
        else:
            self.impuestos_calculados = (base - 18660) * 0.12 + 516

    def generar_declaracion(self):
        return (f"\n========= Declaración =========\n"
                f"Nombre: {self.nombre}\n"
                f"Cédula: {self.cedula}\n"
                f"Dirección: {self.direccion}\n"
                f"Ingresos Totales: ${self.total_ingresos:.2f}\n"
                f"Deducciones Aplicadas: ${self.total_deducciones:.2f}\n"
                f"Impuestos Calculados: ${self.impuestos_calculados:.2f}\n")

File: Solucion_Codigo/ProyectoBim1_ SistemaDeclaracionImpuestos/src/Metodos.java

public class Metodos {

    private String nombre;
    private String cedula;
    private String direccion;

    private double[] sueldosMensuales = new double[12];
    private double[] facturasPorCategoria = new double[6];

    private double totalIngresos;
    private double totalDeducciones;
    private double impuestosCalculados;

    public static final String[] CATEGORIAS = {
        "vivienda", "educacion", "alimentacion", "vestimenta", "salud", "turismo"
    };

    public Metodos(String nombre, String cedula, String direccion) {
        this.nombre = nombre;
        this.cedula = cedula;
        this.direccion = direccion;
    }

    public void agregarSueldo(int mes, double monto) {
        if (mes >= 0 && mes < 12 && monto > 0) {
            sueldosMensuales[mes] = monto;
        }
    }

    public void agregarFactura(double monto, String categoria) {
        int index = obtenerIndiceCategoria(categoria);
        if (index != -1 && monto > 0) {
            facturasPorCategoria[index] += monto;
        }
    }

    private int obtenerIndiceCategoria(String categoria) {
        for (int i = 0; i < CATEGORIAS.length; i++) {
            if (CATEGORIAS[i].equalsIgnoreCase(categoria)) {
                return i;
            }
        }
        return -1;
    }

    public void calcularTotalIngresos() {
        totalIngresos = 0;
        for (double sueldo : sueldosMensuales) {
            totalIngresos += sueldo;
        }
    }

    public void calcularDeducciones() {
        double sumaFacturas = 0;
        for (double monto : facturasPorCategoria) {
            sumaFacturas += monto;
        }
        double limiteDeduccion = Math.min(14754.00, totalIngresos * 0.5);
        totalDeducciones = Math.min(sumaFacturas, limiteDeduccion);
    }

    public void calcularImpuestos() {
        double baseImponible = totalIngresos - totalDeducciones;

        if (baseImponible <= 11722) {
            impuestosCalculados = 0;
        } else if (baseImponible <= 14930) {
            impuestosCalculados = (baseImponible - 11722) * 0.05;
        } else if (baseImponible <= 18660) {
            impuestosCalculados = (baseImponible - 14930) * 0.10 + 160;
        } else {
            impuestosCalculados = (baseImponible - 18660) * 0.12 + 516;
        }
    }

    public String generarDeclaracion() {
        return String.format("""
            ========= Declaracion =========
            Nombre: %s
            Cedula: %s
            Direccion: %s
            Ingresos Totales: $%.2f
            Deducciones Aplicadas: $%.2f
            Impuestos Calculados: $%.2f
            """, nombre, cedula, direccion, totalIngresos, totalDeducciones, impuestosCalculados);
    }
}